proposals/csharp-9.0/records.md

```diff
@@ -197,6 +197,130 @@ If the containing record is abstract, the synthesized clone method is also abstr
 
 
+### Printing members: PrintMembers and ToString
+
+If the record is derived from `object`, the record includes a synthesized method equivalent to a method declared as follows:
+```C#
+protected virtual void PrintMembers(System.StringBuilder builder, bool includeSeparator, bool includeName);
+```
+The method is `virtual` and `protected`.
+It is an error if the method is declared explicitly.
+
+The method:
+1. appends the record name to `builder` followed by " { ", if `includeName` is true
+2. appends a separator ", " to `builder` if `includeSeparator` is true and the record has public fields or properties,
+3. for each of the record's public field and property member, appends that member's name followed by " = " followed by the result of invoking `object.ToString()` on that member's value: `this.member.ToString()`, separated with ", ",
+4. appends " }" if `includeName` is true
+
+If the record type is derived from a base record `Base`, the record includes a synthesized override equivalent to a method declared as follows:
+```C#
+protected override void PrintMembers(StringBuilder builder, bool includeSeparator, bool includeName);
+```
+It is an error if the override is declared explicitly.
+
+The method appends the same contents to `builder` as described above, but calls `base.PrintMembers` between steps 3 and 4 with three arguments:
+1. its `builder` parameter,
+2. a true `includeSeparator` argument if either (a) the record has public fields or properties, or (b) its `includeSeparator` parameter was given as true,
+3. a false `includeName` argument.
+
+The record includes a synthesized method equivalent to a method declared as follows:
+```C#
+public override string ToString();
+```
+
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility, or if the explicit declaration doesn't allow overiding it in a derived type and the record type is not `sealed`. It is an error if either synthesized, or explicitly declared method doesn't override a property with this signature in the record `Base` (for example, if the method is missing in the `Base`, or sealed, or not virtual, etc.).
+
+The synthesized method:
+1. creates a `StringBuilder` instance,
+2. invokes the record's "PrintMembers" method giving it the builder, `includeSeparator` as false, and `includeName` as true
+3. returns the builder's contents with `builder.ToString()`.
+
+For example, consider the following record types:
+
+``` csharp
+record R1(T1 P1);
+record R2(T1 P1, T2 P2, T3 P3) : R1(P1);
+```
+
+For those record types, the synthesized printing members would be something like:
+
+```C#
+class R1
+{
+    public T1 P1 { get; init; }
+    
+    protected virtual PrintMembers(StringBuilder builder, bool includeSeparator, bool includeName)
+    {
+        if (includeName)
+        {
+            builder.Append(nameof(R1));
+            builder.Append(" { ");
+        }
```

Comment:
Can `includeName` be handled directly in `ToString()` in each case?

Reply:
I had it that way in my first draft ;-) I'll change back.